ARROW-8792: [C++][Python][R][GLib] New Array compute kernels implementation and execution framework

wesm opened the pull request:
This patch is a major reworking of our development strategy for implementing array-valued functions and applying them in a query processing setting. 

The design was partly inspired by my previous work designing Ibis (https://github.com/ibis-project/ibis -- the "expr" subsystem and the way that operators validate input types and resolve output types). Using only function names and input types, you can determine the output types of each function and resolve the "execute" function that performs a unit of work processing a batch of data. This will allow us to build deferred column expressions and then (eventually) do parallel execution. 

There are a ton of details, but one nice thing is that there is now a single API entry point for invoking any function by its name:

```c++
Result<Datum> CallFunction(ExecContext* ctx, const std::string& func_name,
                           const std::vector<Datum>& args,
                           const FunctionOptions* options = NULLPTR);
```

What occurs when you do this:

* A `Function` instance is looked up in the global `FunctionRegistry`
* Given the descriptors of `args` (their types and shapes -- array or scalar), the Function searches for `Kernel` that is able to process those types and shapes. A kernel might be able to do `array[T0], array[T1]` or only `scalar[T0], scalar[T1]`, for example. This permits kernel specialization to treat different type and shape combinations
* The kernel is executed iteratively against `args` based on what `args` contains -- if there are ChunkedArrays, they will be split into contiguous pieces. Kernels never see ChunkedArray, only Array or Scalar
* The Executor implementation is able to split contiguous Array inputs into smaller chunks, which is important for parallel execution. See `ExecContext::set_exec_chunksize`

To summarize: the REGISTRY contains FUNCTIONS. A FUNCTION contains KERNELS. A KERNEL is a specific implementation of a function that services a particular type combination. 

An additional effort in this patch is to radically simplify the process of creating kernels that are based on a scalar function. To do this, there is a growing collection of template-based kernel generation classes in compute/kernels/codegen_internal.h that will surely be the topic of much debate. I want to make it a lot easier for people to add new kernels.

There are some other incidental changes in the PR, such as changing the convenience APIs like `Cast` to return `Result`. I'm afraid we may have to live with the API breakage unless someone else wants to add backward compatibility code for the old APIs. 

I have to apologize for making such a large PR. I've been working long hours on this for nearly a month and the process of porting all of our existing functionality and making the unit tests pass caused much iteration in the "framework" part of the code, such that it would have been a huge time drain to review incomplete iterations of the framework that had not been proven to capture the functionality that previously existed in the project.

Given the size of this PR and that fact that it completely blocks any work into src/arrow/compute, I don't think we should let this sit unmerged for more than 4 or 5 days, tops. I'm committed to responding to all of your questions and working to address your feedback about the design and improving the documentation and code comments. I tried to leave copious comments to explain my thought process in various places. Feel free to make any and all comments in this PR or in whatever form you like. I don't think that merging should be blocked on stylistic issues.

wesm commented:
For code reviewers:

* The code diff is a bit deceiving because many kernel files were renamed to make their taxonomy more clear
* I suggest focusing your energies on the files in the arrow/compute top level directory. This is almost entirely new code, what was there before is almost all gone

I am aware that there are some obviously messy things that will need to be cleaned up a bit, but we have to start somewhere.

wesm commented on cpp/CMakeLists.txt:
```diff
@@ -309,6 +309,10 @@ if(ARROW_DATASET)
 
+if(ARROW_GANDIVA)
+  set(ARROW_COMPUTE ON)
+endif()
```
This is unneeded, will remove

wesm commented:
Aside from fixing the builds, I'm going to do a few things to help with this:

* Add as many comments as possible to the new header files
* Add some high level explanation to compute/README.md about how kernels work and how to begin to develop a new one
* Add simple Python bindings for `FunctionRegistry`, `Function`, and `Kernel` so that we can inspect the contents of the registry and the available kernels (and their type signatures) for each function

pitrou commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+};
+
+// TODO: Run benchmarks to determine if OutputAdapter is a zero-cost abstraction
+struct ScalarPrimitiveExec {
```
It's difficult to understand why there's `ScalarPrimitiveExec` as well as `ScalarUnary` and other stuff.

wesm replied:
I will add more comments

wesm replied:
Added a bunch of comments to explain. If we want to create "templates to rule them all" we will need to write sufficient benchmarks to have peace of mind that all the compilers do the right thing

wesm commented on cpp/src/arrow/compute/exec.cc:
```diff
@@ -0,0 +1,932 @@
+
+ protected:
+  Status ExecuteBatch(const ExecBatch& batch, ExecListener* listener) {
+    if (batch.length == 0) {
+      // Skip empty batches. This should only happen with zero-length inputs
+      return Status::OK();
+    }
```
note to self: remove this

wesm commented on cpp/src/arrow/compute/cast.cc:
```diff
@@ -0,0 +1,193 @@
+Type::type CastFunction::out_type_id() const { return impl_->out_type; }
+
+std::unique_ptr<KernelState> CastInit(KernelContext* ctx, const KernelInitArgs& args) {
+  // NOTE: TakeOptions are currently unused, but we pass it through anyway
```
incorrect comment

wesm commented on cpp/src/arrow/compute/exec_internal.h:
```diff
@@ -0,0 +1,137 @@
+                                                        const FunctionOptions* options);
+};
+
+ARROW_EXPORT
+Status ExecuteFunction(ExecContext* ctx, const std::string& func_name,
+                       const std::vector<Datum>& args, const FunctionOptions* options,
+                       ValueDescr* out_descr, ExecListener* listener);
```
This function no longer exists

pitrou commented on cpp/src/arrow/compute/kernels/CMakeLists.txt:
```diff
@@ -15,37 +15,41 @@
+# add_arrow_benchmark(vector_sort_benchmark PREFIX "arrow-compute")
+# add_arrow_benchmark(vector_partition_benchmark PREFIX "arrow-compute")
+# add_arrow_benchmark(vector_filter_benchmark PREFIX "arrow-compute")a
+# add_arrow_benchmark(vector_take_benchmark PREFIX "arrow-compute")
```
Should the benchmarks be re-enabled?

wesm replied:
yes

wesm commented on cpp/src/arrow/compute/kernel.h:
```diff
@@ -15,295 +15,517 @@
+              KernelInit init = NULLPTR)
+      : Kernel(std::move(in_types), std::move(out_type), init), exec(exec) {}
+
+  /// \brief Perform a single invocation of this kernel. In general, this
+  /// function must
```
complete

wesm commented on cpp/src/arrow/compute/kernels/CMakeLists.txt:
```diff
@@ -15,37 +15,41 @@
 # specific language governing permissions and limitations
 # under the License.
```
TODO: clean up this module and compile all benchmarks

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+void BinaryExecFlipped(KernelContext* ctx, ArrayKernelExec exec,
+                       const ExecBatch& batch, Datum* out);
+
+// ----------------------------------------------------------------------
+// Boolean data utilities
+
```
cruft

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+
+namespace codegen {
+
+struct SimpleExec {
```
add comment

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+};
+
+// TODO: Run benchmarks to determine if OutputAdapter is a zero-cost abstraction
+struct ScalarPrimitiveExec {
```
add comments

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+};
+
+template <typename OutType, typename Arg0Type, typename Op>
+struct ScalarUnary {
```
comments

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+// Applies a scalar operation with state on the null-null values of a single
+// array
+template <typename OutType, typename Arg0Type, typename Op>
+struct ScalarUnaryNotNullStateful {
```
more comments

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+
+template <typename OutType, typename Arg0Type, typename Arg1Type, typename Op,
+          typename FlippedOp = Op>
+struct ScalarBinary {
```
more comments

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+};
+
+template <typename OutType, typename Arg0Type, typename Op>
+struct ScalarUnaryNotNull {
```
more comments

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+          typename FlippedOp = Op>
+using ScalarBinaryEqualTypes = ScalarBinary<OutType, ArgType, ArgType, Op, FlippedOp>;
+
+struct ScalarNumericEqualTypes {
```
comments to explain what's going on from here through end of the file

wesm commented on cpp/src/arrow/compute/kernels/scalar_cast.cc:
```diff
@@ -15,37 +15,40 @@
 // specific language governing permissions and limitations
```
remove this file

pitrou commented on cpp/src/arrow/compute/function.h:
```diff
@@ -0,0 +1,203 @@
+  /// \brief Return the first kernel that can execute the function given the
+  /// exact argument types (without implicit type casts or scalar->array
+  /// promotions)
+  virtual Result<const ScalarKernel*> DispatchExact(
```
Why `virtual`? Are there subclasses?

wesm replied:
yes, `CastFunction` has some additional logic

pitrou commented on cpp/src/arrow/compute/kernel.cc:
```diff
@@ -0,0 +1,308 @@
+#include "arrow/compute/exec.h"
+#include "arrow/result.h"
+#include "arrow/util/bit_util.h"
+#include "arrow/util/hashing.h"
```
Required? This pulls a lot of stuff.

wesm replied:
This is only one .cc file, I only need `hash_combine` so perhaps we need `arrow/util/hash_util.h`

pitrou commented on cpp/src/arrow/compute/kernel.h:
```diff
@@ -15,295 +15,517 @@
+  // require single-threaded execution.
+  bool parallelizable = true;
+
+  SimdLevel::type simd_level = SimdLevel::NONE;
```
Is this actually used?

wesm replied:
It will be

wesm commented on cpp/src/arrow/compute/options.h:
```diff
@@ -0,0 +1,155 @@
+
+  // Type being casted to. May be passed separate to eager function
+  // compute::Cast
+  std::shared_ptr<DataType> to_type;
```
Yes, I'm aware, but I wasn't able to figure out another way to communicate the output type to the standard kernel APIs

wesm commented on cpp/src/arrow/compute/options.h:
```diff
@@ -0,0 +1,155 @@
+
+struct ARROW_EXPORT FunctionOptions {};
+
+struct ARROW_EXPORT CastOptions : public FunctionOptions {
```
I don't want to have N headers files where N is the number of different kernels that have options

pitrou replied:
My point is those header files already exist :-) For example, put `CastOptions` in the header file that exposes the `Cast` functions.

wesm replied:
Yes, I realized that after writing, I'll do that and put the other options in the `api_*.h` files

wesm commented on cpp/src/arrow/compute/kernels/scalar_cast_internal.h:
```diff
@@ -0,0 +1,283 @@
+template <typename OutType>
+void AddCommonCasts(OutputType out_ty, CastFunction* func) {
+  // From null to this type
+  DCHECK_OK(func->AddKernel(Type::NA, {InputType::Array(null())}, out_ty,
```
These failures would indicate a failure to initialize the built-in registry, which indicates that the library is broken. If you'd really like to propagate these errors at registry-init time I suggest we pursue this refactoring in follow up work

wesm replied:
I opened https://issues.apache.org/jira/browse/ARROW-8897

wesm commented on cpp/src/arrow/compute/kernels/scalar_cast_internal.h:
```diff
@@ -0,0 +1,283 @@
+    T, enable_if_t<!is_boolean_type<T>::value && !is_nested_type<T>::value &&
+                   !std::is_same<DictionaryType, T>::value>> {
+  static void Add(const OutputType& out_ty, CastFunction* func) {
+    // Dictionary unpacking not implemented for boolean or nested types
```
Yes, that would be better and make the code simpler

wesm replied:
I opened https://issues.apache.org/jira/browse/ARROW-8894

wesm commented on cpp/src/arrow/compute/kernels/scalar_boolean.cc:
```diff
@@ -0,0 +1,189 @@
+
+namespace {
+
+enum class ResolveNull { KLEENE_LOGIC, PROPAGATE };
```
Will look

wesm commented on cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:
```diff
@@ -0,0 +1,52 @@
+    DCHECK_OK(func->AddKernel({InputType::Array(ty), InputType::Array(ty)}, ty,
+                              ScalarNumericEqualTypes::Binary<Op>(*ty)));
+  }
+  DCHECK_OK(registry->AddFunction(std::move(func)));
```
I am not sure I agree but we can deal with this later

wesm replied:
I opened a JIRA https://issues.apache.org/jira/browse/ARROW-8897

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.cc:
```diff
@@ -0,0 +1,153 @@
+  ExecBatch flipped_batch = batch;
+  Datum tmp = flipped_batch.values[0];
+  flipped_batch.values[0] = flipped_batch.values[1];
+  flipped_batch.values[1] = tmp;
```
will do

nealrichardson commented:
The R failure:

```
  ── 1. Error: filter() on timestamp columns (@test-dataset.R#381)  ──────────────
  NotImplemented: Function >= has no kernel matching input types (array[timestamp[us, tz=UTC]], scalar[timestamp[us, tz=UTC]])
  Backtrace:
    1. arrow:::expect_equivalent(...)
    2. dplyr::filter(., ts >= lubridate::ymd_hms("2015-05-04 03:12:39"))
    2. dplyr::filter(., part == 1)
    2. dplyr::select(., ts)
   10. dplyr::collect(.)
   13. Scanner$create(x)$ToTable()
   16. arrow:::dataset___Scanner__ToTable(self)
```

I'm happy to help debug but that sounds pretty straightforward.

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.cc:
```diff
@@ -0,0 +1,153 @@
+}
+
+const std::vector<std::shared_ptr<DataType>>& SignedIntTypes() {
+  std::call_once(codegen_static_initialized, InitStaticData);
```
Will look

pitrou replied:
Just pointing it out for future instances of this pattern.

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+
+#ifdef ARROW_EXTRA_ERROR_CONTEXT
+
+#define KERNEL_ABORT_IF_ERROR(ctx, expr)                \
```
Will fix

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+  int64_t position = 0;
+  typename TypeTraits<Type>::ArrayType arr;
+  ArrayIterator(const ArrayData& data)
+      : arr(data.Copy()) {}
```
Array subclasses require a `shared_ptr<ArrayData>`

pitrou replied:
Oh, right.

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+
+template <typename Type>
+struct GetValueType<Type, enable_if_has_c_type<Type>> {
+  using T = typename Type::c_type;
```
It may not be, I will look

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,648 @@
+  }
+};
+
+// Applies a scalar operation with state on the null-null values of a single
```
yes

wesm commented:
@nealrichardson yeah, I will fix

fsaintjacques commented on cpp/src/arrow/compute/registry.cc:
```diff
@@ -0,0 +1,127 @@
+    return Status::OK();
+  }
+
+  Result<std::shared_ptr<const Function>> GetFunction(const std::string& name) const {
```
You need a lock for every operation, I suggest to use a shared_lock for read operations and exclusive lock for `AddFn..`.

kou commented:
> Port GLib bindings

Can I start working on this on this branch?

wesm commented:
@kou yes please go ahead

kou commented:
OK!

wesm commented:
I rebased without pushing any new changes

emkornfield commented on cpp/src/arrow/compute/api_scalar.cc:
```diff
@@ -0,0 +1,114 @@
+      func_name = "<=";
+      break;
+    default:
+      DCHECK(false);
```
I think you removed default the compiler will warn on uncovered cases.

emkornfield commented:
CC @brills I'm not sure if this impacts TFX at all.

wesm commented:
@emkornfield @brills no Python APIs are changed (not sure how much of the C++ API TFX uses)

emkornfield commented on cpp/src/arrow/compute/exec.h:
```diff
@@ -0,0 +1,160 @@
+struct FunctionOptions;
+class FunctionRegistry;
+
+// It seems like 64K might be a good default chunksize to use for execution
```
Would we want something that fits in L1 cache for most data types?

wesm replied:
Perhaps. We'll have to investigate with benchmarks. Neither parallelization nor operator pipelining is implemented (though this is in my critical path for June) so that will be a good time to do some empirical research.

wesm replied:
I opened https://issues.apache.org/jira/browse/ARROW-8898

emkornfield commented on cpp/src/arrow/compute/exec.h:
```diff
@@ -0,0 +1,160 @@
+  // TODO: At some point we might want the limit the size of contiguous
+  // preallocations (for example, merging small ChunkedArray chunks until
+  // reaching some desired size)
+  void set_preallocate_contiguous(bool preallocate = true) {
```
seems strange to have a default value here.

wesm replied:
fixing

emkornfield commented on cpp/src/arrow/compute/exec.h:
```diff
@@ -0,0 +1,160 @@
+// TODO: Consider standardizing on uint16 selection vectors and only use them
+// when we can ensure that each value is 64K length or smaller
+
+/// \brief Container for a int32 selection
```
it isn't really clear to me what this means.

wesm replied:
adding comments

emkornfield commented on cpp/src/arrow/compute/exec.h:
```diff
@@ -0,0 +1,160 @@
+  explicit SelectionVector(const Array& arr) : SelectionVector(arr.data()) {}
+
+  /// \brief Create SelectionVector from boolean mask
+  static Result<std::shared_ptr<SelectionVector>> FromMask(const Array& arr);
```
should this be BooleanArray?

wesm replied:
Sure. It's not implemented yet

emkornfield commented on cpp/src/arrow/compute/exec.h:
```diff
@@ -0,0 +1,160 @@
+  const int32_t* indices_;
+};
+
+struct ExecBatch {
```
please document.

wesm replied:
done

emkornfield commented on cpp/src/arrow/compute/exec_internal.h:
```diff
@@ -0,0 +1,137 @@
+  /// on the chunk layout of ChunkedArray. Default of -1 means no maximum, so
+  /// as greedy as possible
+  static Result<std::unique_ptr<ExecBatchIterator>> Make(std::vector<Datum> args,
+                                                         int64_t max_chunksize = -1);
```
why not int64_t max default?

wesm replied:
sounds fine, changing

emkornfield commented on cpp/src/arrow/compute/exec_internal.h:
```diff
@@ -0,0 +1,137 @@
+  std::vector<Datum> values_;
+};
+
+Status CheckAllValues(const std::vector<Datum>& values);
```
documentations.

emkornfield commented on cpp/src/arrow/compute/function.h:
```diff
@@ -0,0 +1,203 @@
+
+/// \brief Contains the number of required arguments for the function
+struct ARROW_EXPORT FunctionArity {
+  static FunctionArity Nullary() { return FunctionArity(0, false); }
```
nit: Void or Null for name?

wesm replied:
Hmm https://en.wikipedia.org/wiki/Arity

emkornfield commented on cpp/src/arrow/compute/function.h:
```diff
@@ -0,0 +1,203 @@
+  static FunctionArity Unary() { return FunctionArity(1, false); }
+  static FunctionArity Binary() { return FunctionArity(2, false); }
+  static FunctionArity Ternary() { return FunctionArity(3, false); }
+  static FunctionArity Varargs(int min_args = 1) { return FunctionArity(min_args, true); }
```
nit: VarArgs

It also seems like it might be more readable for callers to have to always specify min_args (especially because it isn't clear to me why 1 is the default and not 0 or 2

wesm replied:
You need at least one argument spec for type checking

wesm replied:
I'm changing the name, but will leave the default issue for follow up work

emkornfield commented on cpp/src/arrow/compute/function.h:
```diff
@@ -0,0 +1,203 @@
+  static FunctionArity Ternary() { return FunctionArity(3, false); }
+  static FunctionArity Varargs(int min_args = 1) { return FunctionArity(min_args, true); }
+
+  FunctionArity(int num_args, bool is_varargs = false)  // NOLINT implicit conversion
```
is implicit conversion really necessary here?

wesm replied:
Well, currently we have code like

```
  auto func = std::make_shared<ScalarFunction>(name, /*arity=*/2);
```

Can't say I'm super fond of writing

```
  auto func = std::make_shared<ScalarFunction>(name, FunctionArity::Binary());
```

but I'm willing to change it

wesm replied:
I'm renaming FunctionArity -> Arity and will make this refactor now

wesm replied:
OK that was easy. This is actually fine

wesm commented:
I got sucked into some other things today, I'll focus on getting the CI passing tomorrow so we can hopefully go into the weekend with a green build

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,710 @@
+
+#endif  // ARROW_EXTRA_ERROR_CONTEXT
+
+// A kernel that exposes Call methods that handles iteration over ArrayData
+// inputs itself
```
note to self: fix this

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,710 @@
+void BinaryExecFlipped(KernelContext* ctx, ArrayKernelExec exec,
+                       const ExecBatch& batch, Datum* out);
+
+// ----------------------------------------------------------------------
+// Template kernel exec function generators
```
fix comment

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,710 @@
+  }
+}
+
+const std::vector<std::shared_ptr<DataType>>& BaseBinaryTypes();
```
Explain what these are for

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,710 @@
+  }
+};
+
+template <typename Type, typename Enable = void>
```
Explain

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,710 @@
+//
+// This template executes the operator even on the data behind null values,
+// therefore it is generally only suitable for operators that cannot fail.
+template <typename OutType, typename Arg0Type, typename Op>
```
Add example functor definition

wesm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -0,0 +1,710 @@
+//
+// This template executes the operator even on the data behind null values,
+// therefore it is generally only suitable for operators that cannot fail.
+template <typename OutType, typename Arg0Type, typename Arg1Type, typename Op,
```
Add example functor

wesm commented on cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:
```diff
@@ -0,0 +1,93 @@
+  result->child_data.push_back(casted_child.array());
+}
+
+OutputType kOutputTargetType(ResolveOutputFromOptions);
```
note to self: add static

wesm commented on cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc:
```diff
@@ -0,0 +1,565 @@
+// specific language governing permissions and limitations
+// under the License.
+
+// Implementation of casting to integer or floating point types
```
note also includes decimal

wesm commented:
I just rebased and pushed MSVC build fixes (that works for me at least locally on VS 2017). I'm going to fix the R failure next and then address the other accumulated comments above.

As far as timeline for this, I would ideally like to have this merged on Sunday at the latest. I was planning to take next week mostly off of work so this would unblock PRs into arrow/compute

I also would like in the next couple days to write a small follow up patch or two to provide a template for scalar string kernels to help unblock ARROW-555. I will also start creating an umbrella JIRA issue for a more general kernels library buildout (a plan for going from our ~20 kernels to 100+)

wesm commented:
This problem of kernel dispatch with timestamps that may or may not have time zones actually brought out a limitation with input type checking. I'm introducing a simple interface for user-defined type-checking rules. I think this should give us sufficient flexibility for the time being, and if not we can always improve the interfaces.

wesm commented:
The R test suite is fixed -- this also turned up some missing test coverage for things that R was depending on https://issues.apache.org/jira/browse/ARROW-8895.

I'm going to grind through the pending comments and add the simple Python bindings I talked about above. I think beyond that (and fixing remaining CI issues) the GLib bindings are the main things left. @kou please let me know if I can be of assistance

kou commented:
I'm working on this now.
It'll be finished in a few hours.

wesm commented:
Awesome, thanks @kou!

kou commented:
It seems that `IsIn(const Datum&, std::shared_ptr<ChunkedArray>)` support is removed. Is it expected?

wesm commented:
@kou yes, that was an API change. I can try to restore this functionality though?

wesm commented:
I'm taking a look now, I'll report back in a half hour or so

wesm commented:
I just pushed a fix for an issue that was showing up in some Windows builds and some other cleaning

wesm commented:
The IsIn/Match changes don't look too hard, should be able to push them in the next half hour or less

wesm commented:
@kou done, thank you for catching this, I think it's better to retain this functionality.

kou commented:
Thanks!
I've also done the GLib part.

kou commented:
It seems that the added `TestIsInKernel.IsInChunkedArrayInvoke` test is failed:

```text
[ RUN      ] TestIsInKernel.IsInChunkedArrayInvoke
../src/arrow/testing/gtest_util.cc:91: Failure
Expected equality of these values:
  expected.num_chunks()
    Which is: 2
  actual.num_chunks()
    Which is: 1
# chunks unequal
[  FAILED  ] TestIsInKernel.IsInChunkedArrayInvoke (1 ms)
```

wesm commented:
Thank you, I will fix.

wesm commented:
OK, I have it fixed locally. I'm finishing some Python stuff and then I'll push

wesm commented:
Done. It looks like GitHub is having some problems today but will keep an eye on the builds

wesm commented:
@xhochy @pitrou @kszucs @jorisvandenbossche  I just added Python bindings for this new functionality including generic argument packing and function dispatching. The new implementation of `pyarrow.compute.sum` for example is now

```
def sum(array):
    """
    Sum the values in a numerical (chunked) array.

    Parameters
    ----------
    array : pyarrow.Array or pyarrow.ChunkedArray

    Returns
    -------
    sum : pyarrow.Scalar
    """
    return call_function('sum', [array])
```

So we're going to be able to go around the Python codebase and delete a _ton_ of code. 

I haven't written unit tests for this part (let's do this later), but I added wrappers for the registry, functions, and kernels:

```
In [1]: import pyarrow.compute as pc                                                                      

In [2]: reg = pc.function_registry()                                                                      

In [3]: reg.list_functions()                                                                              
Out[3]: 
['add',
 'and',
 'and_kleene',
 'count',
 'dict_encode',
 'equals',
 'filter',
 'greater',
 'greater_equal',
 'invert',
 'isin',
 'less',
 'less_equal',
 'match',
 'mean',
 'minmax',
 'not_equals',
 'or',
 'or_kleene',
 'partition_indices',
 'sort_indices',
 'sum',
 'take',
 'unique',
 'value_counts',
 'xor']

In [4]: reg.get_function('equals')                                                                        
Out[4]: 
arrow.compute.Function
kind: scalar
num_kernels: 21

In [5]: reg.get_function('equals').list_kernels()                                                         
Out[5]: 
[ScalarKernel<(any[bool], any[bool]) -> bool>,
 ScalarKernel<(any[uint8], any[uint8]) -> bool>,
 ScalarKernel<(any[uint16], any[uint16]) -> bool>,
 ScalarKernel<(any[uint32], any[uint32]) -> bool>,
 ScalarKernel<(any[uint64], any[uint64]) -> bool>,
 ScalarKernel<(any[int8], any[int8]) -> bool>,
 ScalarKernel<(any[int16], any[int16]) -> bool>,
 ScalarKernel<(any[int32], any[int32]) -> bool>,
 ScalarKernel<(any[int64], any[int64]) -> bool>,
 ScalarKernel<(any[float], any[float]) -> bool>,
 ScalarKernel<(any[double], any[double]) -> bool>,
 ScalarKernel<(any[binary], any[binary]) -> bool>,
 ScalarKernel<(any[string], any[string]) -> bool>,
 ScalarKernel<(any[large_binary], any[large_binary]) -> bool>,
 ScalarKernel<(any[large_string], any[large_string]) -> bool>,
 ScalarKernel<(any[date32[day]], any[date32[day]]) -> bool>,
 ScalarKernel<(any[date64[ms]], any[date64[ms]]) -> bool>,
 ScalarKernel<(any[timestamp(s)], any[timestamp(s)]) -> bool>,
 ScalarKernel<(any[timestamp(ms)], any[timestamp(ms)]) -> bool>,
 ScalarKernel<(any[timestamp(us)], any[timestamp(us)]) -> bool>,
 ScalarKernel<(any[timestamp(ns)], any[timestamp(ns)]) -> bool>]
```

Needless to say very excited about all of this.

wesm commented:
I'm just fixing a failing unit test now.

kou commented:
I've fixed a build error with `-DCMAKE_UNITY_BUILD=ON`.

wesm commented:
Perfect thank you :pray:. I'm done hacking on this for now, I'll keep investigating test failures

kou commented:
We need the following change for macOS. What do you thin about this change?

```diff
diff --git a/cpp/src/arrow/compute/kernel.cc b/cpp/src/arrow/compute/kernel.cc
index 1d02117cc..2895468c8 100644
--- a/cpp/src/arrow/compute/kernel.cc
+++ b/cpp/src/arrow/compute/kernel.cc
@@ -158,8 +158,8 @@ std::shared_ptr<TypeMatcher> TimestampUnit(TimeUnit::type unit) {
 // ----------------------------------------------------------------------
 // InputType
 
-uint64_t InputType::Hash() const {
-  uint64_t result = kHashSeed;
+size_t InputType::Hash() const {
+  size_t result = kHashSeed;
   hash_combine(result, static_cast<int>(shape_));
   hash_combine(result, static_cast<int>(kind_));
   switch (kind_) {
@@ -341,11 +341,11 @@ bool KernelSignature::MatchesInputs(const std::vector<ValueDescr>& args) const {
   return true;
 }
 
-uint64_t KernelSignature::Hash() const {
+size_t KernelSignature::Hash() const {
   if (hash_code_ != 0) {
     return hash_code_;
   }
-  uint64_t result = kHashSeed;
+  size_t result = kHashSeed;
   for (const auto& in_type : in_types_) {
     hash_combine(result, in_type.Hash());
   }
diff --git a/cpp/src/arrow/compute/kernel.h b/cpp/src/arrow/compute/kernel.h
index cd59bfa82..f90627ab5 100644
--- a/cpp/src/arrow/compute/kernel.h
+++ b/cpp/src/arrow/compute/kernel.h
@@ -198,7 +198,7 @@ class ARROW_EXPORT InputType {
   bool operator!=(const InputType& other) const { return !(*this == other); }
 
   /// \brief Return hash code
-  uint64_t Hash() const;
+  size_t Hash() const;
 
   /// \brief Render a human-readable string representation
   std::string ToString() const;
@@ -359,7 +359,7 @@ class ARROW_EXPORT KernelSignature {
   bool operator!=(const KernelSignature& other) const { return !(*this == other); }
 
   /// \brief Compute a hash code for the signature
-  uint64_t Hash() const;
+  size_t Hash() const;
 
   const std::vector<InputType>& in_types() const { return in_types_; }
 
```

wesm commented:
@kou that's fine, please go ahead and apply it

wesm commented:
Note that `kHashSeed` is declared as `uint64_t` so that must be changed also

```
static constexpr uint64_t kHashSeed = 0;
```

kou commented:
Done.

kou commented:
All build errors are fixed.

There are some test failures with MSVC:

https://github.com/apache/arrow/pull/7240/checks?check_run_id=701142871

```text
2020-05-23T01:07:32.9299182Z [ RUN      ] TestCallScalarFunction.PreallocationCases
2020-05-23T01:07:32.9299744Z D:/a/arrow/arrow/cpp/src/arrow/compute/exec_test.cc:560:  Check failed: !internal::BitmapEquals(arg0.buffers[0]->data(), arg0.offset, out_arr->buffers[0]->data(), out_arr->offset, batch.length) 
```

wesm commented:
Ok I will take a look

wesm commented:
I see the problem, I am fixing

wesm commented:
The build is passing (the Travis thing is a flake on ARM architecture), thanks @kou for your help with that! I think we should merge this today and I will continue to address comments in this PR in follow up patches. I'll wait until later today but it would be good if others could chime in to make sure this plan is agreeable.

wesm commented:
I'd like to work on a couple follow-up PRs tomorrow, so in order to unblock these PRs so that they can be reviewed, this PR will have to be merged first. So I would like to do that tomorrow absent problems that need to be fixed. 

Upon merging this PR, I would immediately reopen it so that others can continue to use the GitHub code review feature to leave comments in mass, and I will certainly respond to those comments and address them or open follow up JIRA issues so they are tracked. Please also see the "build out" umbrella JIRA where I will track the follow on work building out kernels functionality in the project

https://issues.apache.org/jira/browse/ARROW-8894

jorisvandenbossche commented:
I was playing a little bit with the python API:

- Ran into the following issue:

    ```python
    >>> import pyarrow.compute as pc                                                                                                                                                                               

    # works nicely
    >>> pc.call_function("take", [pa.array([0, 1, 2, 3, 4, 5]), pa.array([0, 4, 1, None])])                                                                                                                        
    <pyarrow.lib.Int64Array object at 0x7f7cc8954a68>
    [
      0,
      4,
      1,
      null
    ]

    # segfaults
    >>> pc.call_function("filter", [pa.array([0, 1, 2, 3]), pa.array([True, False, True, None])])                                                                                                                  
    Segmentation fault (core dumped)
    ```

* In array.pxi / table.pxi, some of the methods' implementation has been replaced with a `_pc.call_function(..)` call, and some are still using the C++ compute functions (eg `Take`). The goal is to eventually have them all go through `_pc.call_function()`?

* `pc.call_function()` doesn't yet have a way to pass options. That's also something to be added in the future?

wesm commented:
@jorisvandenbossche I'll take a look at that segfault

> The goal is to eventually have them all go through _pc.call_function()?

Yes. `arrow::compute::CallFunction` is general enough to deal with non-value (aka not Array or Scalar) arguments so we should be able to define table-valued functions that dispatch array kernels to a table or record batch's columns

> pc.call_function() doesn't yet have a way to pass options. That's also something to be added in the future?

Yes, I'll open a JIRA about this

wesm commented:
@jorisvandenbossche I fixed the segfault you hit and now have changed the implementation of `Array.filter` to be

```
pc = _pc()
options = pc.FilterOptions(null_selection_behavior)
return pc.call_function('filter', [self, mask], options)
```

I also rebased -- I am going to merge this when the build is passing so we do not continue to stack changes that need to be reviewed on this PR. I welcome further code reviews on this PR or follow up JIRAs

wesm commented:
+1

wesm commented:
The PR has been merged, but the code review feature is still available. Please continue to leave comments and I will address them

wesm commented:
I'll leave this PR open until next week sometime to allow more time for comments

wesm commented:
Closing the PR again. Please open JIRA issues or review other PRs to provide more feedback.
